contracts/src/v2/Types.sol

```diff
@@ -0,0 +1,170 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2023 Snowfork <hello@snowfork.com>
+pragma solidity 0.8.28;
+
+import {OperatingMode} from "./../types/Common.sol";
+
+// Inbound message from a Polkadot parachain (via BridgeHub)
+struct InboundMessage {
+    // origin
+    bytes32 origin;
+    // Message nonce
+    uint64 nonce;
+    // Commands
+    Command[] commands;
+}
+
+struct Command {
+    uint8 kind;
+    uint64 gas;
+    bytes payload;
+}
+
+library CommandKind {
+    uint8 constant Upgrade = 0;
+    uint8 constant SetOperatingMode = 1;
+    uint8 constant UnlockNativeToken = 2;
+    uint8 constant RegisterForeignToken = 3;
+    uint8 constant MintForeignToken = 4;
+    uint8 constant CallContract = 5;
+}
+
+struct Payload {
+    // sender of the message
+    address origin;
+    Asset[] assets;
+    Xcm xcm;
+    bytes claimer;
+    // ether value
+    uint128 value;
+    // additional ether value for execution fees
+    uint128 executionFee;
+    // additional ether value for relayer fees
+    uint128 relayerFee;
+}
```

Comment:
I'd suggest adding an `id` field as V1 for tracing/indexing purposes.

The on-chain change is detailed in
https://github.com/paritytech/polkadot-sdk/pull/7402#discussion_r1976823135

Reply:
I do not think we need to do this. I think generating a hash based on the nonce is a good enough solution. However if the UX designs to set the topic id manually in the XCM we should obey it. I have added more details here:
https://github.com/paritytech/polkadot-sdk/pull/7402/files/27189594b6ef9affbda2d2c6f870c67de6fdce3a#r1980442243